File: 2023/domaci/44-matersky-programovaci-jazyk/volfmatous/main.js
const readline = require("readline");
const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout,
    terminal: false,
});

let linesRead = 0;
const lines = [];

async function readLine() {
    if (lines.length < 1) {
        for await (const line of rl) {
            lines.push(line);
        }
    }

    linesRead++;
    return lines[linesRead - 1];
}

main().then();

async function main() {
    const variables = {};

    const c = parseInt(await readLine());

    const linesSplit = [];
    const linesNestLevels = [];

    for (let i = 0; i < c; i++) {
        let line = (await readLine()).split("//")[0].trimEnd();

        if (line === '') {
            continue;
        }

        let nestLevel = 0;
        while (line.length > 2 && line.startsWith("  ")) {
            line = line.substring(2);
            nestLevel++;
        }

        linesSplit.push(line.split(' '));
        linesNestLevels.push(nestLevel);
    }

    const output = [];

    let repeating = false;
    let repeatStartLineIndex = 0;
    let nestLevelCurrent = 0;

    for (let i = 0; i < linesSplit.length; i++) {
        const lineSplit = linesSplit[i];
        const nestLevel = linesNestLevels[i];

        if (nestLevel > nestLevelCurrent) {
            continue;
        }

        const atLastLine = i >= linesSplit.length - 1;

        if (repeating && (nestLevel < linesNestLevels[repeatStartLineIndex] + 1 || atLastLine)) {
            repeating = false;
            i = repeatStartLineIndex - 1;
            if (!atLastLine) {
                continue;
            }
        } else if (nestLevel < nestLevelCurrent) {
            nestLevelCurrent = nestLevel;
        }

        switch (lineSplit[0]) {
            case "kdyz": {
                if (lineSplit.length < 5 || lineSplit[lineSplit.length - 1] !== "tak") {
                    console.log("error");
                    return;
                }

                const conditionResult = conditionPass(lineSplit.slice(1, -1), variables);

                if (!conditionResult[0]) {
                    console.log("error");
                    return;
                }

                if (conditionResult[1]) {
                    nestLevelCurrent++;
                }

                break;
            }

            case "dokud": {
                if (lineSplit.length < 5 || lineSplit[lineSplit.length - 1] !== "tak") {
                    console.log("error");
                    return;
                }

                const conditionResult = conditionPass(lineSplit.slice(1, -1), variables);

                if (!conditionResult[0]) {
                    console.log("error");
                    return;
                }

                if (conditionResult[1]) {
                    nestLevelCurrent++;
                    repeating = true;
                    repeatStartLineIndex = i;
                }

                break;
            }

            case "vypsat": {
                const expressionResult = expressionValue(lineSplit.slice(1), variables);

                if (!expressionResult[0]) {
                    console.log("error");
                    return;
                }

                output.push(expressionResult[1]);
                break;
            }

            default: {
                const variableName = lineSplit[0];

                if (variableName.toLowerCase().charCodeAt(0) < 'a'.charCodeAt(0) || variableName.toLowerCase().charCodeAt(0) > 'z'.charCodeAt(0)
                    || lineSplit.length < 2 || lineSplit[1] !== "je") {
                    console.log("error");
                    return;
                }

                switch (variableName) {
                    case "je":
                    case "a":
                    case "bez":
                    case "kdyz":
                    case "tak":
                    case "rovno":
                    case "aspon":
                    case "nejvic":
                    case "nebo":
                    case "dokud":
                    case "vypsat":
                        console.log("error");
                        return;
                }

                const expressionResult = expressionValue(lineSplit.slice(2), variables);

                if (!expressionResult[0]) {
                    console.log("error");
                    return;
                }

                variables[variableName] = expressionResult[1];
                break;
            }
        }
    }

    for (const line of output) {
        console.log(line);
    }
}

function expressionValue(lineSplit, variables) {
    if (lineSplit.length % 2 !== 1) {
        return [false, 0];
    }

    let expressionAdd = true;
    let expressionValue = 0;

    for (let w = 0; w < lineSplit.length; w++) {
        if (w % 2 === 0) {
            let value = parseFloat(lineSplit[w]);
            value = isNaN(value) ? variables[lineSplit[w]] : value;
            if (value === undefined) {
                return [false, 0];
            }

            expressionValue += expressionAdd ? value : -value;
        } else {
            switch (lineSplit[w]) {
                case "a":
                    expressionAdd = true;
                    break;
                case "bez":
                    expressionAdd = false;
                    break;
                default:
                    return [false, 0];
            }
        }
    }

    return [true, expressionValue];
}

function conditionPass(lineSplit, variables) {
    const orIndexes = [-1];
    for (let i = 0; i < lineSplit.length; i++) {
        if (lineSplit[i] === "nebo") {
            orIndexes.push(i);
        }
    }

    orIndexes.push(lineSplit.length);

    let conditionPass = false;

    for (let i = 0; i < orIndexes.length - 1; i++) {
        const startIndex = orIndexes[i] + 1;
        const endIndex = orIndexes[i + 1] - 1;

        let comparatorIndex = -1;
        for (let c = startIndex; c <= endIndex; c++) {
            if (lineSplit[c] === "rovno" || lineSplit[c] === "aspon" || lineSplit[c] === "nejvic") {
                comparatorIndex = c;
                break;
            }
        }

        if (comparatorIndex === -1) {
            return [false, false];
        }

        const comparator = lineSplit[comparatorIndex];
        const expression1Result = expressionValue(lineSplit.slice(startIndex, comparatorIndex), variables);
        const expression2Result = expressionValue(lineSplit.slice(comparatorIndex + 1, endIndex + 1), variables);

        if (!expression1Result[0] || !expression2Result[0]) {
            return [false, false];
        }

        if (!conditionPass) {
            const expression1 = expression1Result[1];
            const expression2 = expression2Result[1];

            switch (comparator) {
                case "rovno":
                    conditionPass = expression1 === expression2;
                    break;
                case "aspon":
                    conditionPass = expression1 >= expression2;
                    break;
                case "nejvic":
                    conditionPass = expression1 <= expression2;
                    break;
            }
        }
    }

    return [true, conditionPass];
}

File: 2023/domaci/44-matersky-programovaci-jazyk/volfmatous/main.py
import sys
from typing import Dict, List, Tuple, Union


def expression_value(line_split: List[str], variables: Dict[str, Union[int, float]]) -> Tuple[bool, Union[int, float]]:
    if len(line_split) % 2 != 1:
        return False, 0

    expression_add = True
    expression_value = 0

    for w in range(len(line_split)):
        if w % 2 == 0:
            value = float(line_split[w]) if line_split[w].replace('.', '').replace('-', '').isdigit() else variables.get(line_split[w], None)

            if value is None:
                return False, 0

            expression_value += value if expression_add else -value
        else:
            if line_split[w] == "a":
                expression_add = True
            elif line_split[w] == "bez":
                expression_add = False
            else:
                return False, 0

    return True, expression_value


def condition_pass(line_split: List[str], variables: Dict[str, Union[int, float]]) -> Tuple[bool, bool]:
    or_indexes = [-1]
    for i in range(len(line_split)):
        if line_split[i] == "nebo":
            or_indexes.append(i)
    or_indexes.append(len(line_split))

    condition_pass = False

    for i in range(len(or_indexes) - 1):
        start_index = or_indexes[i] + 1
        end_index = or_indexes[i + 1] - 1

        comparator_index = -1
        for c in range(start_index, end_index + 1):
            if line_split[c] in ["rovno", "aspon", "nejvic"]:
                comparator_index = c
                break

        if comparator_index == -1:
            return False, False

        comparator = line_split[comparator_index]
        expression1_result = expression_value(line_split[start_index:comparator_index], variables)
        expression2_result = expression_value(line_split[comparator_index + 1:end_index + 1], variables)

        if not expression1_result[0] or not expression2_result[0]:
            return False, False

        if not condition_pass:
            expression1 = expression1_result[1]
            expression2 = expression2_result[1]

            if comparator == "rovno":
                condition_pass = expression1 == expression2
            elif comparator == "aspon":
                condition_pass = expression1 >= expression2
            elif comparator == "nejvic":
                condition_pass = expression1 <= expression2

    return True, condition_pass

if __name__ == "__main__":
    variables = {}

    c = int(input())

    lines_split = []
    lines_nest_levels = []

    for _ in range(c):
        line = input().split("//")[0].rstrip()

        if line == '':
            continue

        nest_level = 0
        while len(line) > 2 and line.startswith("  "):
            line = line[2:]
            nest_level += 1

        lines_split.append(line.split(' '))
        lines_nest_levels.append(nest_level)

    output = []

    repeating = False
    repeat_start_line_index = 0
    nest_level_current = 0

    i = 0
    while i < len(lines_split):
        line_split = lines_split[i]
        nest_level = lines_nest_levels[i]

        if nest_level > nest_level_current:
            i += 1
            continue

        at_last_line = i >= len(lines_split) - 1

        if repeating and (nest_level < lines_nest_levels[repeat_start_line_index] + 1 or at_last_line):
            repeating = False
            i = repeat_start_line_index - 1
            if not at_last_line:
                i += 1
                continue
        elif nest_level < nest_level_current:
            nest_level_current = nest_level

        if line_split[0] == "kdyz":
            if len(line_split) < 5 or line_split[-1] != "tak":
                print("error")
                sys.exit()

            condition_result = condition_pass(line_split[1:-1], variables)

            if not condition_result[0]:
                print("error")
                sys.exit()

            if condition_result[1]:
                nest_level_current += 1
        elif line_split[0] == "dokud":
            if len(line_split) < 5 or line_split[-1] != "tak":
                print("error")
                sys.exit()

            condition_result = condition_pass(line_split[1:-1], variables)

            if not condition_result[0]:
                print("error")
                sys.exit()

            if condition_result[1]:
                nest_level_current += 1
                repeating = True
                repeat_start_line_index = i
        elif line_split[0] == "vypsat":
            expression_result = expression_value(line_split[1:], variables)

            if not expression_result[0]:
                print("error")
                sys.exit()

            output.append(expression_result[1])
        else:
            variable_name = line_split[0]

            if not 'a' <= variable_name[0].lower() <= 'z' or len(line_split) < 2 or line_split[1] != "je":
                print("error")
                sys.exit()

            if variable_name in ["je", "a", "bez", "kdyz", "tak", "rovno", "aspon", "nejvic", "nebo", "dokud", "vypsat"]:
                print("error")
                sys.exit()

            expression_result = expression_value(line_split[2:], variables)

            if not expression_result[0]:
                print("error")
                sys.exit()

            variables[variable_name] = expression_result[1]

        i += 1

    for line in output:
        print(f"{line:g}")
